src/main/java/org/folio/rest/impl/InventoryHelper.java

```diff
@@ -0,0 +1,429 @@
+
+  private final HttpClientInterface httpClient;
+  private final Map<String, String> okapiHeaders;
+  private final Context ctx;
+  private final String lang;
+
+  InventoryHelper(HttpClientInterface httpClient, Map<String, String> okapiHeaders, Context ctx, String lang) {
+    this.httpClient = httpClient;
+    this.okapiHeaders = okapiHeaders;
+    this.ctx = ctx;
+    this.lang = lang;
+  }
+
+  public CompletableFuture<PoLine> handleInstanceRecord(PoLine compPOL) {
+    return getProductTypesMap(compPOL)
+      .thenCompose(productTypesMap -> getInstanceRecord(compPOL, productTypesMap))
+      .thenApply(compPOL::withInstanceId);
+  }
+
+  public CompletableFuture<String> handleHoldingRecord(PoLine compPOL) {
+    // TODO the logic here will be implemented in scope of MODORDERS-66
+    String endpoint = "/holdings-storage/holdings?limit=1";
+    return handleGetRequest(endpoint, httpClient, ctx, okapiHeaders, logger)
+      .thenApply(response -> getFirstObjectFromResponse(response, "holdingsRecords"))
+      .thenApply(this::extractId);
+  }
+
+  /**
+   * Returns list of item id's corresponding to given PO line.
+   * Items are either retrieved from Inventory or new ones are created if no corresponding item records exist yet.
+   *
+   * @param compPOL   PO line to retrieve/create Item Records for
+   * @param holdingId holding uuid from the inventory
+   * @return future with list of item id's
+   */
+  public CompletableFuture<List<String>> handleItemRecords(PoLine compPOL, String holdingId) {
+    int expectedCount = calculateInventoryItemsQuantity(compPOL);
+    return searchForExistingItems(compPOL)
+      .thenCompose(existingItemIds -> createMissingItems(compPOL, holdingId, expectedCount - existingItemIds.size())
+        .thenApply(createdItemIds -> ListUtils.union(existingItemIds, createdItemIds))
+      );
+  }
+
+  /**
+   * Retrieves product type details associated with given PO line
+   * and builds 'product type name' -> 'product type id' map.
+   *
+   * @param compPOL the PO line to retrieve product type details for
+   * @return product types map
+   */
+  private CompletableFuture<Map<String, String>> getProductTypesMap(PoLine compPOL) {
+    // do not fail if no productId is provided, should be enforced on schema level if it's required
+    if (compPOL.getDetails() == null || compPOL.getDetails().getProductIds().isEmpty()) {
+      return completedFuture(Collections.emptyMap());
+    }
+
+    String endpoint = compPOL.getDetails().getProductIds().stream()
+      .map(productId -> String.format("name==%s", productId.getProductIdType().toString()))
+      .collect(joining(" or ", "/identifier-types?query=", ""));
+
+    return handleGetRequest(endpoint, httpClient, ctx, okapiHeaders, logger)
+      .thenApply(productTypes -> {
+        if (productTypes.getJsonArray("identifierTypes").size() != compPOL.getDetails().getProductIds().size()) {
+          throw new CompletionException(new HttpException(422,
```

Comment:
I personally think this is a misuse of the HttpException class, which was intended to encapsulate an HTTP 4xx/5xx response.  Here, we never got a 422, with this message.  While this may be convenient, it could be confusing to someone debugging a problem in logs, etc.   

I would prefer to see a new Exception class introduced for this specific purpose, e.g. `InvalidProductTypeException`